packages/cli-repl/src/cli-repl.ts

```diff
@@ -78,10 +77,10 @@ class CliRepl {
    * @param {NodeOptions} driverOptions - The driver options.
    */
   async setupRepl(driverUri: string, driverOptions: NodeOptions): Promise<void> {
-    this.serviceProvider = await this.connect(driverUri, driverOptions);
-    this.ShellEvaluator = new ShellEvaluator(this.serviceProvider, this.bus, this);
-    this.buildInfo = await this.serviceProvider.buildInfo();
-    this.logBuildInfo(driverUri);
+    const initialServiceProvider = await this.connect(driverUri, driverOptions);
+    this.internalState = new ShellInternalState(initialServiceProvider, this.bus);
+    this.shellEvaluator = new ShellEvaluator(this.internalState, this);
+    await this.internalState.fetchConnectionInfo();
```

Comment:
Wouldn't be possible for this one to return a `ConnectionInfo` object and cache  it rather than mutate this.internalState?

Reply:
It may be that on initial connect we just keep this call entirely in internalState instead of awaiting it here.

Because if the connection information changes, technically the internal state also has to change. I think it really ought to mutate the state (and then log that change).

Reply:
Phrased this one wrong: totally get the mutation has to happen internally. It seems safe to me to keep that 'private' inside `ShellInternalState`.

Reply:
My thinking is that the connection info is internal state, therefore it should be stored in the internal state object. It is exactly the type of information that needs to be used and updated in various places, so why hide it? 

I don't see the benefit to caching the connection info in the repl when we want that information to be stored in a place where it can be updated dynamically (so for example, when the user reassigns the `db` variable in the shell, the internal state class will automatically update it's connection info object). The repl's prompt should be based on that internal state connection information, so when we implement custom prompts then we will get that info from the internal state. Perhaps we will have the repls pass a callback to the internal state so it can update the prompt when the `db` changes, but since we haven't implemented that feature and it's not a part of this refactor, I left that unimplemented.

`fetchConnectionInfo` is a separate async function because it needs to do a DB command to get that info, and in the browser-repl the place where the ShellInternalState is instantiated isn't an async function.